### PR TITLE
[BM-609] Add pairing devices list

### PR DESCRIPTION
**Pull request by gabrysgab**
### Task

 - [JIRA](https://netguru.atlassian.net/browse/BM-609)
 
### Description

Handling pairing devices by showing a list instead of automatically connecting to the first device.

### CODE REVIEW

File: app/src/main/kotlin/co/netguru/baby/monitor/client/common/extensions/ViewExtensions.kt [hunk 1]
```diff
@@ -107,3 +110,19 @@ fun <T> ImageView.babyProfileImage(uri: T, borderSize: Float, colorRes: Int, vec
             }
         })
 }
```
Comment (netguru-hound) [inline app/src/main/kotlin/co/netguru/baby/monitor/client/common/extensions/ViewExtensions.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/netguru/baby-monitor-client-android/blob/76162f8000da5effbb33db8436698272b4874b84/app/src/main/kotlin/co/netguru/baby/monitor/client/common/extensions/ViewExtensions.kt#L112"></span>Needless blank line(s)</td>
    </tr>
  </tbody>
</table>

File: app/src/main/kotlin/co/netguru/baby/monitor/client/feature/client/configuration/ConnectingDevicesFragment.kt [hunk 2]
```diff
@@ -1,51 +1,108 @@
...
+            }
+            is NsdState.Completed -> {
+                handleServices(nsdState.serviceInfoList)
+                setupRefreshButton()
+            }
         }
     }
 
+    private fun handleNsdServiceError(throwable: Throwable) {
+        when(throwable) {
```
Comment (netguru-hound) [inline app/src/main/kotlin/co/netguru/baby/monitor/client/feature/client/configuration/ConnectingDevicesFragment.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/netguru/baby-monitor-client-android/blob/76162f8000da5effbb33db8436698272b4874b84/app/src/main/kotlin/co/netguru/baby/monitor/client/feature/client/configuration/ConnectingDevicesFragment.kt#L78"></span>Missing spacing after “when”</td>
    </tr>
  </tbody>
</table>

File: app/src/main/kotlin/co/netguru/baby/monitor/client/feature/client/configuration/ConnectingDevicesFragment.kt [hunk 3]
```diff
@@ -1,51 +1,108 @@
...
+            is ResolveFailedException ->  showSnackbarMessage(R.string.discovering_services_error)
```
Comment (netguru-hound) [inline app/src/main/kotlin/co/netguru/baby/monitor/client/feature/client/configuration/ConnectingDevicesFragment.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/netguru/baby-monitor-client-android/blob/76162f8000da5effbb33db8436698272b4874b84/app/src/main/kotlin/co/netguru/baby/monitor/client/feature/client/configuration/ConnectingDevicesFragment.kt#L79"></span>Unnecessary space(s)</td>
    </tr>
  </tbody>
</table>

File: app/src/main/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/DeviceNameProvider.kt [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package co.netguru.baby.monitor.client.feature.communication.nsd
+
+import com.jaredrummler.android.device.DeviceName
+import javax.inject.Inject
+
+class DeviceNameProvider @Inject constructor(): IDeviceNameProvider {
```
Comment (netguru-hound) [inline app/src/main/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/DeviceNameProvider.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/netguru/baby-monitor-client-android/blob/76162f8000da5effbb33db8436698272b4874b84/app/src/main/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/DeviceNameProvider.kt#L6"></span>Missing spacing before “:”</td>
    </tr>
  </tbody>
</table>

File: app/src/main/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/NsdServiceExceptions.kt [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package co.netguru.baby.monitor.client.feature.communication.nsd
+
+import java.lang.RuntimeException
+
+class ResolveFailedException: RuntimeException("Service resolution failed")
```
Comment (netguru-hound) [inline app/src/main/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/NsdServiceExceptions.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/netguru/baby-monitor-client-android/blob/76162f8000da5effbb33db8436698272b4874b84/app/src/main/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/NsdServiceExceptions.kt#L5"></span>Missing spacing before “:”</td>
    </tr>
  </tbody>
</table>

File: app/src/main/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/NsdServiceExceptions.kt [hunk 6]
```diff
@@ -0,0 +1,7 @@
...
+class RegistrationFailedExcetpion: RuntimeException("Service registration failed")
```
Comment (netguru-hound) [inline app/src/main/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/NsdServiceExceptions.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/netguru/baby-monitor-client-android/blob/76162f8000da5effbb33db8436698272b4874b84/app/src/main/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/NsdServiceExceptions.kt#L6"></span>Missing spacing before “:”</td>
    </tr>
  </tbody>
</table>

File: app/src/main/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/NsdServiceExceptions.kt [hunk 7]
```diff
@@ -0,0 +1,7 @@
...
+class StartDiscoveryFailedException: RuntimeException("Service discovery failed")
```
Comment (netguru-hound) [inline app/src/main/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/NsdServiceExceptions.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/netguru/baby-monitor-client-android/blob/76162f8000da5effbb33db8436698272b4874b84/app/src/main/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/NsdServiceExceptions.kt#L7"></span>Missing spacing before “:”</td>
    </tr>
  </tbody>
</table>

File: app/src/main/kotlin/co/netguru/baby/monitor/client/feature/settings/ConfigurationViewModel.kt [hunk 8]
```diff
@@ -35,19 +37,12 @@ class ConfigurationViewModel @Inject constructor(
...
-                handleNewService(service.host.hostAddress, service.port)
-            }
-        }
-    }
+    val nsdStateLiveData: LiveData<NsdState> = nsdServiceManager.nsdStateLiveData
 
-    private fun handleNewService(
-        address: String,
-        port: Int
+    fun handleNewService(
```
Comment (netguru-hound) [inline app/src/main/kotlin/co/netguru/baby/monitor/client/feature/settings/ConfigurationViewModel.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/netguru/baby-monitor-client-android/blob/76162f8000da5effbb33db8436698272b4874b84/app/src/main/kotlin/co/netguru/baby/monitor/client/feature/settings/ConfigurationViewModel.kt#L42"></span>Unexpected indentation (expected 8, actual 7)</td>
    </tr>
  </tbody>
</table>

File: app/src/test/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/NsdServiceManagerTest.kt [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    @Test
+    fun `should post error state on StartDiscoveryFailedException`() {
+        nsdServiceManager.discoverService()
+
+        verify(nsdManager).discoverServices(any(), any(), argumentCaptorDiscovery.capture())
+        argumentCaptorDiscovery.firstValue.onStartDiscoveryFailed("", 0)
+
+        verify(nsdStateObserver).onChanged(argThat { this is NsdState.Error && throwable is StartDiscoveryFailedException })
```
Comment (netguru-hound) [inline app/src/test/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/NsdServiceManagerTest.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/netguru/baby-monitor-client-android/blob/76162f8000da5effbb33db8436698272b4874b84/app/src/test/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/NsdServiceManagerTest.kt#L86"></span>Line detected that is longer than the defined maximum line length in the code style.</td>
    </tr>
  </tbody>
</table>

File: app/src/test/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/NsdServiceManagerTest.kt [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+    @Test
+    fun `should post error state on ResolveFailedException`() {
+        nsdServiceManager.discoverService()
+
+        verify(nsdManager).discoverServices(any(), any(), argumentCaptorDiscovery.capture())
+        argumentCaptorDiscovery.firstValue.onServiceFound(nsdServiceInfo)
+
+        verify(nsdManager).resolveService(eq(nsdServiceInfo), argumentCaptorResolve.capture())
+        argumentCaptorResolve.firstValue.onResolveFailed(nsdServiceInfo ,0)
```
Comment (netguru-hound) [inline app/src/test/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/NsdServiceManagerTest.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/netguru/baby-monitor-client-android/blob/76162f8000da5effbb33db8436698272b4874b84/app/src/test/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/NsdServiceManagerTest.kt#L97"></span>Unexpected spacing before “,”</td>
    </tr>
  </tbody>
</table>

Comment (netguru-hound) [inline app/src/test/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/NsdServiceManagerTest.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/netguru/baby-monitor-client-android/blob/76162f8000da5effbb33db8436698272b4874b84/app/src/test/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/NsdServiceManagerTest.kt#L97"></span>Missing spacing after “,”</td>
    </tr>
  </tbody>
</table>

**Comment by netguru-hound** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          13 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/netguru/baby-monitor-client-android/blob/76162f8000da5effbb33db8436698272b4874b84/app/src/main/res/values/dimens.xml#L3">app/src/main/res/values/dimens.xml#L3</a> - The resource <code>R.dimen.text_size_xxsmall</code> appears to be unused</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/netguru/baby-monitor-client-android/blob/76162f8000da5effbb33db8436698272b4874b84/app/src/main/res/values/dimens.xml#L8">app/src/main/res/values/dimens.xml#L8</a> - The resource <code>R.dimen.text_size_big</code> appears to be unused</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/netguru/baby-monitor-client-android/blob/76162f8000da5effbb33db8436698272b4874b84/app/src/main/res/values/dimens.xml#L10">app/src/main/res/values/dimens.xml#L10</a> - The resource <code>R.dimen.text_size_xxlarge</code> appears to be unused</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/netguru/baby-monitor-client-android/blob/76162f8000da5effbb33db8436698272b4874b84/app/src/main/res/values/dimens.xml#L11">app/src/main/res/values/dimens.xml#L11</a> - The resource <code>R.dimen.text_size_huge</code> appears to be unused</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/netguru/baby-monitor-client-android/blob/76162f8000da5effbb33db8436698272b4874b84/app/src/main/res/values/dimens.xml#L15">app/src/main/res/values/dimens.xml#L15</a> - The resource <code>R.dimen.spacing_xsmall</code> appears to be unused</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/netguru/baby-monitor-client-android/blob/76162f8000da5effbb33db8436698272b4874b84/app/src/main/res/values/dimens.xml#L24">app/src/main/res/values/dimens.xml#L24</a> - The resource <code>R.dimen.spacing_xnormal</code> appears to be unused</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/netguru/baby-monitor-client-android/blob/76162f8000da5effbb33db8436698272b4874b84/app/src/main/res/values/dimens.xml#L26">app/src/main/res/values/dimens.xml#L26</a> - The resource <code>R.dimen.spacing_xxxnormal</code> appears to be unused</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/netguru/baby-monitor-client-android/blob/76162f8000da5effbb33db8436698272b4874b84/app/src/main/res/values/dimens.xml#L29">app/src/main/res/values/dimens.xml#L29</a> - The resource <code>R.dimen.spacing_xlarge</code> appears to be unused</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/netguru/baby-monitor-client-android/blob/76162f8000da5effbb33db8436698272b4874b84/app/src/main/res/values/dimens.xml#L30">app/src/main/res/values/dimens.xml#L30</a> - The resource <code>R.dimen.spacing_xxlarge</code> appears to be unused</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/netguru/baby-monitor-client-android/blob/76162f8000da5effbb33db8436698272b4874b84/app/src/main/res/values/dimens.xml#L35">app/src/main/res/values/dimens.xml#L35</a> - The resource <code>R.dimen.toolbar_elevation</code> appears to be unused</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/netguru/baby-monitor-client-android/blob/76162f8000da5effbb33db8436698272b4874b84/app/src/main/res/values/dimens.xml#L39">app/src/main/res/values/dimens.xml#L39</a> - The resource <code>R.dimen.padding_normal</code> appears to be unused</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/netguru/baby-monitor-client-android/blob/76162f8000da5effbb33db8436698272b4874b84/app/src/main/res/values/dimens.xml#L41">app/src/main/res/values/dimens.xml#L41</a> - The resource <code>R.dimen.shadow_padding</code> appears to be unused</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/netguru/baby-monitor-client-android/blob/76162f8000da5effbb33db8436698272b4874b84/app/src/main/res/values/strings.xml#L71">app/src/main/res/values/strings.xml#L71</a> - The resource <code>R.string.discovery_start_failed</code> appears to be unused</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">:link: <a href="https://netguru.atlassian.net/browse/BM-609">BM-609</a></td>
    </tr>
  </tbody>
</table>

### JaCoCO Code Coverage 11.65% :white_check_mark:
| Class | Covered | Meta | Status |
|:---|:---:|:---:|:---:|
| `co/netguru/baby/monitor/client/feature/client/home/SendBabyNameUseCase` | 100% | 0% | :white_check_mark: |
| `co/netguru/baby/monitor/client/feature/settings/ConfigurationViewModel` | 50% | 0% | :white_check_mark: |
| `co/netguru/baby/monitor/client/application/ApplicationModule` | 0% | 0% | :white_check_mark: |
| `co/netguru/baby/monitor/client/feature/communication/nsd/DeviceNameProvider` | 0% | 0% | :white_check_mark: |
| `co/netguru/baby/monitor/client/feature/communication/nsd/ServiceViewHolder` | 0% | 0% | :white_check_mark: |
| `co/netguru/baby/monitor/client/feature/communication/nsd/NsdServiceDiffUtil` | 0% | 0% | :white_check_mark: |
| `co/netguru/baby/monitor/client/feature/communication/nsd/NsdServiceManager` | 40% | 0% | :white_check_mark: |
| `co/netguru/baby/monitor/client/feature/communication/nsd/NsdServicesAdapter` | 0% | 0% | :white_check_mark: |
| `co/netguru/baby/monitor/client/feature/server/ServerViewModel` | 64% | 0% | :white_check_mark: |
| `co/netguru/baby/monitor/client/feature/server/ChildMonitorFragment` | 0% | 0% | :white_check_mark: |
| `co/netguru/baby/monitor/client/feature/client/configuration/ConnectingDevicesFragment` | 0% | 0% | :white_check_mark: |


#### APK file size

| size |
| --- |
| 30MB |



#### Number of method references

| Dex file | count |
| --- | --- |
| classes.dex | 40470 |



<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: app/src/test/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/NsdServiceManagerTest.kt [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+    @Test
+    fun `should post error state on StartDiscoveryFailedException`() {
+        nsdServiceManager.discoverService()
+
+        verify(nsdManager).discoverServices(any(), any(), argumentCaptorDiscovery.capture())
+        argumentCaptorDiscovery.firstValue.onStartDiscoveryFailed("", 0)
+
+        verify(nsdStateObserver).onChanged(argThat {
+            this is NsdState.Error
+                    && throwable is StartDiscoveryFailedException
```
Comment (netguru-hound) [inline app/src/test/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/NsdServiceManagerTest.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/netguru/baby-monitor-client-android/blob/d115bca3285b89631e71206e29f5d0ceecee0375/app/src/test/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/NsdServiceManagerTest.kt#L88"></span>Line must not begin with “&amp;&amp;”</td>
    </tr>
  </tbody>
</table>

File: app/src/test/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/NsdServiceManagerTest.kt [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+
+        verify(nsdManager).discoverServices(any(), any(), argumentCaptorDiscovery.capture())
+        argumentCaptorDiscovery.firstValue.onServiceFound(nsdServiceInfo)
+
+        verify(nsdManager).resolveService(eq(nsdServiceInfo), argumentCaptorResolve.capture())
+        argumentCaptorResolve.firstValue.onResolveFailed(nsdServiceInfo, 0)
+
+        verify(nsdStateObserver).onChanged(argThat {
+            this is NsdState.Error
+                    && throwable is ResolveFailedException
```
Comment (netguru-hound) [inline app/src/test/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/NsdServiceManagerTest.kt]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/netguru/baby-monitor-client-android/blob/d115bca3285b89631e71206e29f5d0ceecee0375/app/src/test/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/NsdServiceManagerTest.kt#L104"></span>Line must not begin with “&amp;&amp;”</td>
    </tr>
  </tbody>
</table>

File: app/src/main/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/NsdServicesAdapter.kt [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+            ViewGroup.LayoutParams.MATCH_PARENT,
+            view.resources.getDimension(R.dimen.nsd_service_item_height).toInt()
+        )
+        view.layoutParams = layoutParams
+        return ServiceViewHolder(view)
+    }
+
+    override fun onBindViewHolder(holder: ServiceViewHolder, position: Int) {
+        holder.bind(getItem(position))
+        holder.itemView.setOnClickListener {
```
Comment (Gralls) [inline app/src/main/kotlin/co/netguru/baby/monitor/client/feature/communication/nsd/NsdServicesAdapter.kt]: Better approach is to add onClickListeners in `onCreateViewHolder`. It'll not be attached to every single item in recycler.

**Pull request merged**